source/server/hot_restart_impl.cc

```diff
@@ -105,18 +112,25 @@ std::string SharedMemory::version(size_t max_num_stats, size_t max_stat_name_len
 }
 
 std::string SharedMemory::version() {
-  return version(num_stats_, Stats::RawStatData::maxNameLength());
+  return version(max_stats_, Stats::RawStatData::maxNameLength());
 }
 
 HotRestartImpl::HotRestartImpl(Options& options)
-    : options_(options), shmem_(SharedMemory::initialize(options)), log_lock_(shmem_.log_lock_),
+    : options_(options), stats_set_(sharedMemHashOptions(options)),
+      stats_set_data_size_(stats_set_.numBytes()),
+      shmem_(SharedMemory::initialize(stats_set_data_size_, options)), log_lock_(shmem_.log_lock_),
       access_log_lock_(shmem_.access_log_lock_), stat_lock_(shmem_.stat_lock_),
       init_lock_(shmem_.init_lock_) {
   my_domain_socket_ = bindDomainSocket(options.restartEpoch());
   child_address_ = createDomainSocketAddress((options.restartEpoch() + 1));
   initDomainSocketAddress(&parent_address_);
+  bool need_init = true;
   if (options.restartEpoch() != 0) {
     parent_address_ = createDomainSocketAddress((options.restartEpoch() + -1));
+    need_init = !stats_set_.attach(shmem_.stats_set_data_, stats_set_data_size_);
```

Comment:
OK, I switched to init/attach during construction, throwing on error.

Of Envoy's variances with the Google Style Guide, the ok-to-throw policy is the hardest for me to get used to, but I understand the motivation and I'll get used to it.

I chose to throw rather than RELEASE_ASSERT so I could test it.  I'm not sure if I'm doing everything right (e.g. I'm catching in the test but not in production; is that the right thing?) and I just put a string in to describe what I'm throwing, as opposed to making a new std::exception object.

Reply:
I think you can actually test RELEASE_ASSERT.  Look for death tests in the codebase.  But that being said, I think it's ok to throw.

Almost every thrown exception is an instance of EnvoyException.  So unless you need to be able to catch specific exceptions by type, just use that.